### PR TITLE
Multisubject dataset

### PR DESCRIPTION
**Pull request by EmmaRenauld**
## Description
The next big step is underway. Reminder:
1. Hdf5 data creation: done
2. Training: 1) Changing arg parser for yaml file (done)  2) TrainerAbstract class:
   A. multidataset  ---> THIS IS THE CURRENT STEP
   B. batch sampler: will depend on the model.
3. Tracking
4. Testing

In this PR:
- single subject datasets:
-- Instead of having the lazy version overriding the non-lazy, creation of an abstract parent class for both the SubjectMRIData (which contains 1 group of data) and SubjectData (which contains the streamlines + all groups). Note that I also modified the data creation to add attrs 'type' to indicate what king of data we are currently loading (MRI-like or streamlines). This will allow for easier modification if someone wants to add different types of input to the data.
--Modifiying the mri data to be a list of mri data instead, one per group in the groups_config file that we integrated in the hdf creation. 
- multi subject datasets:
-- The collate_fn is actually quite model-dependant: Here, one x (volume), one target (streamlines). Moved to the file batch_sampler, and moved that file inside model: DO NOT REVIEW, this will be for a next step.
-- Also moved all parameters that had no influence on the loading (lazy vs non-lazy) of data, but are rather linked to the way we sample a batch (ex, add noise, reverse streamlines, even neighborhood).
-- Once again, instead of having the lazy version overriding the non-lazy, creation of an abstract parent class and adding groups management from the groups_config.
- data_lists:
-- Same, adding parent class.

To do in next PRs: 
-- using SFT instead of nib.streamlines.ArraySequence for the streamlines?

## Testing data and script
See the new "tests" folder. I added tests for the multisubject_containers, and an example of how to use the test script.

## Have you
- [x ] Added a description of the content of this PR above
- [ ] Followed [proper commit message formatting](https://chris.beams.io/posts/git-commit/)
- [ x] Added data and a script to test this PR
- [ ] Made sure that PEP8 issues are resolved
- [x ] Tested the code yourself right before pushing
- [ ] Added unit tests to test the code you implemented

### CODE REVIEW

File: dwi_ml/model/batch_samplers.py [hunk 1]
```diff
@@ -166,3 +184,370 @@ def __iter__(self):
...
+            previous_dirs = [torch.cat((torch.zeros((1, 3),
+                                                    dtype=torch.float32,
+                                                    device=device),
+                                        d[:-1]))
+                             for d in batch_directions]
+            batch_x_data = [torch.cat((s, p), dim=1)
+                            for s, p in zip(batch_x_data, previous_dirs)]
+
+        return batch_x_data
+>>>>>>> 4eb53f0... Batch_sampler now in model. Separated model-dependant functions from multisubject class and into batch_sampler. Arranged multisubject to include groups from config file.
```
Comment (ppoulin91) [inline dwi_ml/model/batch_samplers.py:553]: On dirait un log qui a été copié-collé, c'est normal?

File: setup.py [hunk 2]
```diff
@@ -44,5 +44,5 @@
     package_data={},
     data_files=[],
     entry_points={},
-    scripts=glob.glob("scripts_python/*.py")
+    scripts=glob.glob("scripts_python/*.py") + glob.glob("tests/*.py")
```
Comment (ppoulin91) [inline setup.py:47]: On veut vraiment exposer les scripts de tests lors de l'installation?

File: tests/test_multisubjectdataset_creation_from_hdf5.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import logging
+import os
+from os import path
+from argparse import RawTextHelpFormatter
+
+import numpy as np
+from dipy.io.stateful_tractogram import Origin, Space, StatefulTractogram
+from dipy.io.streamline import save_tractogram
+from nibabel import Nifti1Image
+#from torch.utils.data.dataloader import DataLoader
```
Comment (ppoulin91) [inline tests/test_multisubjectdataset_creation_from_hdf5.py:12]: Imports cleanup

**Review by ppoulin91 — COMMENTED** [review]
Pas beaucoup de commentaires, je te fais confiance pour la majorité des trucs, je ne vais pas tester localement, je vais attendre qu'on ait une version initiale complète

**Review by ppoulin91 — APPROVED** [review]

File: setup.py (hunk 2 above)
Comment (EmmaRenauld) [reply setup.py]: Non, ok, c'est pas obligé

File: dwi_ml/model/batch_samplers.py (hunk 1 above)
Comment (EmmaRenauld) [reply dwi_ml/model/batch_samplers.py:553]: oups

**Pull request merged**